src/response/error.rs

```diff
@@ -1,70 +1,142 @@
+        source: ErrorSource,
+    },
+
     /// Requested operation did not complete in allotted time.
-    Timeout(String),
+    #[error("operation timed out: {description}")]
+    Timeout {
+        description: String,
+        #[source]
+        source: ErrorSource,
+    },
+
     /// Requested operation couldn't be completed because system limits were hit.
-    LimitExceeded(String),
+    #[error("a limit was exceeded: {description}")]
+    LimitExceeded {
+        description: String,
+        #[source]
+        source: ErrorSource,
+    },
+
     /// Represents all client side exceptions thrown by the SDK.
     /// his exception typically implies that the request wasn't sent to the service successfully or if the service responded, the sdk couldn't interpret the response.
     /// An example would be SDK client was unable to convert the user provided data into a valid request that was expected by the service.
-    ClientSdkError(String),
+    #[error("client error: {description}")]
+    ClientSdkError {
+        description: String,
+        #[source]
+        source: ErrorSource,
+    },
+
     /// SDK client side validation fails.
-    InvalidArgument(String),
-}
+    #[error("invalid argument: {description}")]
+    InvalidArgument {
+        description: String,
+        #[source]
+        source: Option<ErrorSource>,
+    },
 
-impl std::fmt::Display for MomentoError {
-    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-        match self {
-            MomentoError::InternalServerError(e)
-            | MomentoError::BadRequest(e)
-            | MomentoError::PermissionDenied(e)
-            | MomentoError::Unauthenticated(e)
-            | MomentoError::NotFound(e)
-            | MomentoError::AlreadyExists(e)
-            | MomentoError::Cancelled(e)
-            | MomentoError::Timeout(e)
-            | MomentoError::LimitExceeded(e)
-            | MomentoError::ClientSdkError(e)
-            | MomentoError::InvalidArgument(e) => write!(f, "{e}"),
-        }
-    }
+    /// Requested operation was interrupted.
+    /// This may happen to a Topic subscription, or it may show up due to an HTTP2 GOAWAY graceful reconnection request from the server.
+    /// Whatever the case, you can probably retry the request.
+    #[error("operation interrupted: {description}")]
+    Interrupted {
+        description: String,
+        #[source]
+        source: ErrorSource,
+    },
 }
 
-impl std::error::Error for MomentoError {}
+#[derive(Debug, thiserror::Error)]
+pub enum ErrorSource {
+    /// A source you will need to downcast if you need to do something with it.
+    #[error("unknown source")]
+    Unknown(#[from] Box<dyn std::error::Error + Send + Sync>),
 
-impl From<http::uri::InvalidUri> for MomentoError {
-    fn from(e: http::uri::InvalidUri) -> Self {
-        // the uri gets derived from the jwt
-        Self::ClientSdkError(e.to_string())
-    }
-}
+    /// Caused by something in our backing library Tonic
```

Comment:
This comment seems inaccurate. `AuthError` is from the `auth` module.

Reply:
mispaste, thanks